wallets/core/src/builders/action.ts

```diff
@@ -0,0 +1,86 @@
+import type { Actions, Context, Operators } from '../hub/namespaces/types.js';
+import type { AnyFunction, FunctionWithContext } from '../types/actions.js';
+
+export interface ActionByBuilder<T, Context> {
+  actionName: keyof T;
+  and: Operators<T>;
+  or: Operators<T>;
+  after: Operators<T>;
+  before: Operators<T>;
+  action: FunctionWithContext<T[keyof T], Context>;
+}
+
+/*
+ * TODO:
+ * Currently, to use this builder you will write something like this:
+ * new ActionBuilder<EvmActions, 'disconnect'>('disconnect').after(....)
+ *
+ * I couldn't figure it out to be able typescript infer the constructor value as key of actions.
+ * Ideal usage:
+ * new ActionBuilder<EvmActions>('disconnect').after(....)
+ *
+ */
+export class ActionBuilder<T extends Actions<T>, K extends keyof T> {
+  readonly name: K;
+  #and: Operators<T> = new Map();
+  #or: Operators<T> = new Map();
+  #after: Operators<T> = new Map();
+  #before: Operators<T> = new Map();
+  #action: FunctionWithContext<T[keyof T], Context<T>> | undefined;
+
+  constructor(name: K) {
+    this.name = name;
+  }
```

Comment:
Could you remove the second generic parameter from the `ActionBuilder` class and replace it with `keyof T` wherever the `K` parameter is used? I'm having trouble understanding why the second generic parameter is necessary

Reply:
It's kinda workaround. First I tried to implement something like this:

```ts
new ActionBuilder<EvmActions>('disconnect').after(....)
```
But I couldn't infer `disconnect` from constructor somehow. we need tp to pick only one action (e.g. `disconnect`) from actions `T` (e.g. `EvmActions`).

So I had to add second param to manually set what is the action. it's now like:
```ts
new ActionBuilder<EvmActions, 'disconnect'>('disconnect').after(....)
```